File: src/main/java/com/hp/jipp/encoding/EnumType.java
package com.hp.jipp.encoding;

import com.google.auto.value.AutoValue;
import com.google.common.base.Function;
import com.google.common.base.Optional;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Lists;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.Collection;
import java.util.List;
import java.util.Map;

public class EnumType<T extends NameCode> extends AttributeType<T> {

    /** Create a new enumeration encoder */
    public static <T extends NameCode> Encoder<T> encoder(String name, Collection<T> enums,
            NameCode.Factory<T> factory) {
        return new AutoValue_EnumType_Encoder<>(name,
                new ImmutableMap.Builder<Integer, T>().putAll(NameCode.toMap(enums)).build(),
                factory);
    }

    /**
     * An encoder for NameCode enumerations. Use {@link #encoder(String, Collection, NameCode.Factory)}
     * to create instance for new EnumTypes.
     */
    @AutoValue
    public abstract static class Encoder<T extends NameCode> extends com.hp.jipp.encoding.Encoder<T> {

        /** Return the user-visible name of the enum (for debugging purposes) */
        public abstract String getName();

        /** Return the map all known enums */
        public abstract Map<Integer, T> getEnumsMap();

        /** Return a factory for constructing new enum instances */
        abstract NameCode.Factory<T> getFactory();

        /** Returns a known enum, or creates a new instance if not found */
        public T getEnum(int code) {
            Optional<T> e = Optional.fromNullable(getEnumsMap().get(code));
            if (e.isPresent()) return e.get();
            return getFactory().create(getName() + "(x" + Integer.toHexString(code) + ")", code);
        }

        @Override
        T readValue(DataInputStream in, Tag valueTag) throws IOException {
            expectLength(in, 4);
            return getEnum(in.readInt());
        }

        @Override
        void writeValue(DataOutputStream out, T value) throws IOException {
            out.writeShort(4);
            out.writeInt(value.getCode());
        }

        @Override
        boolean valid(Tag valueTag) {
            return valueTag == Tag.EnumValue;
        }
    }

    private final Function<Integer, T> toEnum = new Function<Integer, T>() {
        @Override
        public T apply(Integer input) {
            return ((Encoder<T>)getEncoder()).getEnum(input);
        }
    };

    public EnumType(Encoder<T> encoder, String name) {
        super(encoder, Tag.EnumValue, name);
    }

    @Override
    @SuppressWarnings({"unchecked"})
    public Optional<Attribute<T>> from(Attribute<?> attribute) {
        if (attribute.getValueTag() != Tag.EnumValue) {
            return Optional.absent();
        }
        return Optional.of(of(Lists.transform((List<Integer>)attribute.getValues(), toEnum)));
    }
}

File: src/main/java/com/hp/jipp/encoding/EnumType.kt
// Copyright 2017 HP Development Company, L.P.
// SPDX-License-Identifier: MIT

package com.hp.jipp.encoding

import com.hp.jipp.util.getStaticObjects
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

/** An IPP enum type */
open class EnumType<T : Enum>(val enumEncoder: EnumType.Encoder<T>, override val name: String) :
        AttributeType<T>(enumEncoder, Tag.enumValue) {

    /**
     * An [Encoder] for [Enum] values
     * @param typeName Human-readable type of the [Enum]
     * @param map predefined [Enum] instances to reuse when decoding
     * @param factory a way to create new [Enum] instances of the correct type when decoding an undefined value
     */
    data class Encoder<T : Enum>(
        override val typeName: String,
        val map: Map<Int, T>,
        val factory: (code: Int, name: String) -> T
    ) : com.hp.jipp.encoding.Encoder<T>() {
        constructor(name: String, enums: Collection<T>, factory: (code: Int, name: String) -> T):
                this(name, Enum.toCodeMap(enums), factory)

        /** Returns a known [Enum], or creates a new instance from factory if not found  */
        operator fun get(code: Int): T =
            map[code] ?: factory(code, "$typeName(x${Integer.toHexString(code)})")

        @Throws(IOException::class)
        override fun readValue(input: DataInputStream, finder: Finder, valueTag: Tag): T {
            return get(IntegerType.Encoder.readValue(input, valueTag))
        }

        @Throws(IOException::class)
        override fun writeValue(out: DataOutputStream, value: T) {
            IntegerType.Encoder.writeValue(out, value.code)
        }

        override fun valid(valueTag: Tag): Boolean = valueTag == Tag.enumValue
    }

    override fun of(attribute: Attribute<*>): Attribute<T>? =
        if (attribute.valueTag !== Tag.enumValue) null
        else of(attribute.values
                .filter { it is Int }
                .map { enumEncoder[it as Int] })
}

/**
 * Create an [EnumType] encoder for a subclass of Enum. All public static instances of the class
 * will be included as potential values for decoding purposes.
 */
fun <T : Enum> encoderOf(cls: Class<T>, factory: (code: Int, name: String) -> T) =
        cls.run {
            EnumType.Encoder(simpleName, getStaticObjects()
                    .filter { isAssignableFrom(it.javaClass) }
                    .map {
                        @Suppress("UNCHECKED_CAST")
                        it as T
                    }, factory)
        }

/**
 * Create an [EnumType] encoder for a subclass of Enum. All public static instances of the class
 * will be included as potential values for decoding purposes.
 */
fun <T : Enum> encoderOf(cls: Class<T>, factory: Enum.Factory<T>) =
        encoderOf(cls) { code, name -> factory.of(code, name) }
